h/views/badge.py

```diff
@@ -38,7 +51,14 @@ def badge(request):
     if not uri:
         raise httpexceptions.HTTPBadRequest()
 
-    if models.Blocklist.is_blocked(request.db, uri):
+    # Do a cheap check to see if this URI has ever been annotated. If not,
+    # and most haven't, then we can skip the costs of a blocklist lookup or
+    # search request. In addition to the Elasticsearch query, the search request
+    # involves several DB queries to expand URIs and enumerate group IDs
+    # readable by the current user.
+    if not _has_uri_ever_been_annotated(request.db, uri):
```

Comment:
This might read better as `if _uri_has_never_been_annotated(...):` (i.e. reverse the polarity of the helper function!)

Reply:
Hmm, I tried this but found the function itself slightly confusing to reason about then because of the way the SQL query is naturally written (`SELECT EXISTS (...)` returns a single row with true if a match is found or false otherwise).